// @colyseus/sdk/build/transport/H3Transport.cjs
// Copyright (c) 2026 Endel Dreyer.
//
// This software is released under the MIT License.
// https://opensource.org/license/MIT
//
// colyseus.js@0.17.42
'use strict';

var tslib = require('tslib');
var schema = require('@colyseus/schema');

// 9 bytes is the maximum length of a variable-length integer prefix
const MAX_LENGTH_PREFIX_BYTES = 9;
/**
 * Reassembles length-prefixed frames from arbitrary byte chunks.
 *
 * A single WebTransport `reader.read()` may:
 *   - deliver multiple whole frames in one chunk
 *   - split a frame (or its length prefix) across multiple chunks
 *
 * This reassembler buffers partial data across reads so each dispatched
 * frame is exactly one complete message.
 */
class FrameReassembler {
    constructor() {
        this.pending = new Uint8Array(0);
    }
    push(chunk) {
        if (!chunk || chunk.byteLength === 0) {
            return [];
        }
        const bytes = (this.pending.byteLength === 0)
            ? chunk
            : concatBytes(this.pending, chunk);
        const frames = [];
        let offset = 0;
        while (offset < bytes.byteLength) {
            const it = { offset };
            let length;
            try {
                length = schema.decode.number(bytes, it);
            }
            catch (e) {
                // length prefix is incomplete — wait for more bytes
                if (bytes.byteLength - offset <= MAX_LENGTH_PREFIX_BYTES) {
                    break;
                }
                throw e;
            }
            const frameEnd = it.offset + length;
            if (frameEnd > bytes.byteLength) {
                // payload is incomplete — wait for more bytes
                break;
            }
            frames.push(bytes.subarray(it.offset, frameEnd));
            offset = frameEnd;
        }
        this.pending = (offset < bytes.byteLength)
            ? bytes.slice(offset)
            : new Uint8Array(0);
        return frames;
    }
}
function concatBytes(a, b) {
    const out = new Uint8Array(a.byteLength + b.byteLength);
    out.set(a, 0);
    out.set(b, a.byteLength);
    return out;
}
class H3TransportTransport {
    constructor(events) {
        this.isOpen = false;
        this.lengthPrefixBuffer = new Uint8Array(9); // 9 bytes is the maximum length of a length prefix
        this.reliableReassembler = new FrameReassembler();
        this.unreliableReassembler = new FrameReassembler();
        this.events = events;
    }
    connect(url, options = {}) {
        const wtOpts = options.fingerprint && ({
            // requireUnreliable: true,
            // congestionControl: "default", // "low-latency" || "throughput"
            serverCertificateHashes: [{
                    algorithm: 'sha-256',
                    value: new Uint8Array(options.fingerprint).buffer
                }]
        }) || undefined;
        this.wt = new WebTransport(url, wtOpts);
        this.wt.ready.then((e) => {
            console.log("WebTransport ready!", e);
            this.isOpen = true;
            this.unreliableReader = this.wt.datagrams.readable.getReader();
            this.unreliableWriter = this.wt.datagrams.writable.getWriter();
            const incomingBidi = this.wt.incomingBidirectionalStreams.getReader();
            incomingBidi.read().then((stream) => {
                this.reader = stream.value.readable.getReader();
                this.writer = stream.value.writable.getWriter();
                // immediately write room/sessionId for establishing the room connection
                this.sendSeatReservation(options.roomId, options.sessionId, options.reconnectionToken, options.skipHandshake);
                // start reading incoming data
                this.readIncomingData();
                this.readIncomingUnreliableData();
            }).catch((e) => {
                console.error("failed to read incoming stream", e);
                console.error("TODO: close the connection");
            });
            // this.events.onopen(e);
        }).catch((e) => {
            // this.events.onerror(e);
            // this.events.onclose({ code: e.closeCode, reason: e.reason });
            console.log("WebTransport not ready!", e);
            this._close();
        });
        this.wt.closed.then((e) => {
            console.log("WebTransport closed w/ success", e);
            this.events.onclose({ code: e.closeCode, reason: e.reason });
        }).catch((e) => {
            console.log("WebTransport closed w/ error", e);
            this.events.onerror(e);
            this.events.onclose({ code: e.closeCode, reason: e.reason });
        }).finally(() => {
            this._close();
        });
    }
    send(data) {
        const prefixLength = schema.encode.number(this.lengthPrefixBuffer, data.length, { offset: 0 });
        const dataWithPrefixedLength = new Uint8Array(prefixLength + data.length);
        dataWithPrefixedLength.set(this.lengthPrefixBuffer.subarray(0, prefixLength), 0);
        dataWithPrefixedLength.set(data, prefixLength);
        this.writer.write(dataWithPrefixedLength);
    }
    sendUnreliable(data) {
        const prefixLength = schema.encode.number(this.lengthPrefixBuffer, data.length, { offset: 0 });
        const dataWithPrefixedLength = new Uint8Array(prefixLength + data.length);
        dataWithPrefixedLength.set(this.lengthPrefixBuffer.subarray(0, prefixLength), 0);
        dataWithPrefixedLength.set(data, prefixLength);
        this.unreliableWriter.write(dataWithPrefixedLength);
    }
    close(code, reason) {
        try {
            this.wt.close({ closeCode: code, reason: reason });
        }
        catch (e) {
            console.error(e);
        }
    }
    readIncomingData() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            let result;
            while (this.isOpen) {
                try {
                    result = yield this.reader.read();
                    //
                    // a single read may contain multiple messages
                    // each message is prefixed with its length
                    // a read may also deliver a partial frame; buffer across reads
                    //
                    for (const frame of this.reliableReassembler.push(result.value)) {
                        this.events.onmessage({ data: frame });
                    }
                }
                catch (e) {
                    if (e.message.indexOf("session is closed") === -1) {
                        console.error("H3Transport: failed to read incoming data", e);
                    }
                    break;
                }
                if (result.done) {
                    break;
                }
            }
        });
    }
    readIncomingUnreliableData() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            let result;
            while (this.isOpen) {
                try {
                    result = yield this.unreliableReader.read();
                    //
                    // a single read may contain multiple messages
                    // each message is prefixed with its length
                    // a read may also deliver a partial frame; buffer across reads
                    //
                    for (const frame of this.unreliableReassembler.push(result.value)) {
                        this.events.onmessage({ data: frame });
                    }
                }
                catch (e) {
                    if (e.message.indexOf("session is closed") === -1) {
                        console.error("H3Transport: failed to read incoming data", e);
                    }
                    break;
                }
                if (result.done) {
                    break;
                }
            }
        });
    }
    sendSeatReservation(roomId, sessionId, reconnectionToken, skipHandshake) {
        const it = { offset: 0 };
        const bytes = [];
        schema.encode.string(bytes, roomId, it);
        schema.encode.string(bytes, sessionId, it);
        if (reconnectionToken) {
            schema.encode.string(bytes, reconnectionToken, it);
        }
        if (skipHandshake) {
            schema.encode.boolean(bytes, 1, it);
        }
        this.writer.write(new Uint8Array(bytes).buffer);
    }
    _close() {
        this.isOpen = false;
    }
}

exports.FrameReassembler = FrameReassembler;
exports.H3TransportTransport = H3TransportTransport;
