// @colyseus/sdk/build/fetchXHR.cjs
// Copyright (c) 2026 Endel Dreyer.
//
// This software is released under the MIT License.
// https://opensource.org/license/MIT
//
// colyseus.js@0.17.42
'use strict';

var tslib = require('tslib');

/**
 * Minimal fetch-compatible wrapper around XMLHttpRequest.
 * Used as an automatic fallback when globalThis.fetch is unavailable
 * (e.g. Cocos Creator Native).
 */
function xhrFetch(url, init) {
    return new Promise((resolve, reject) => {
        var _a;
        const xhr = new XMLHttpRequest();
        const method = (init === null || init === void 0 ? void 0 : init.method) || "GET";
        xhr.open(method, url.toString());
        xhr.withCredentials = ((init === null || init === void 0 ? void 0 : init.credentials) === "include");
        // Apply request headers
        if (init === null || init === void 0 ? void 0 : init.headers) {
            const headers = (init.headers instanceof Headers)
                ? init.headers
                : new Headers(init.headers);
            headers.forEach((value, key) => {
                xhr.setRequestHeader(key, value);
            });
        }
        xhr.onload = () => {
            var _a;
            // Parse response headers
            const headers = new Headers();
            const rawHeaders = xhr.getAllResponseHeaders().trim();
            if (rawHeaders) {
                for (const line of rawHeaders.split(/[\r\n]+/)) {
                    const idx = line.indexOf(": ");
                    if (idx > 0) {
                        headers.append(line.substring(0, idx), line.substring(idx + 2));
                    }
                }
            }
            const responseBody = (_a = xhr.response) !== null && _a !== void 0 ? _a : xhr.responseText;
            resolve(new XHRResponse(responseBody, {
                status: xhr.status,
                statusText: xhr.statusText,
                headers,
            }));
        };
        xhr.onerror = () => reject(new TypeError("Network request failed"));
        xhr.ontimeout = () => reject(new TypeError("Network request timed out"));
        xhr.send((_a = init === null || init === void 0 ? void 0 : init.body) !== null && _a !== void 0 ? _a : null);
    });
}
/**
 * Minimal Response-compatible class backed by XHR response data.
 * Implements only the surface used by HTTP.executeRequest().
 */
class XHRResponse {
    constructor(body, init) {
        this.body = body;
        this.status = init.status;
        this.statusText = init.statusText;
        this.headers = init.headers;
        this.ok = init.status >= 200 && init.status < 300;
    }
    json() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            return typeof this.body === "string"
                ? JSON.parse(this.body)
                : this.body;
        });
    }
    text() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            return typeof this.body === "string"
                ? this.body
                : JSON.stringify(this.body);
        });
    }
    blob() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            return new Blob([this.body]);
        });
    }
}

exports.xhrFetch = xhrFetch;
