// @colyseus/sdk/build/index.cjs
// Copyright (c) 2026 Endel Dreyer.
//
// This software is released under the MIT License.
// https://opensource.org/license/MIT
//
// colyseus.js@0.17.42
'use strict';

require('./legacy.cjs');
var Client = require('./Client.cjs');
var Room = require('./Room.cjs');
var Auth = require('./Auth.cjs');
var Errors = require('./errors/Errors.cjs');
var sharedTypes = require('@colyseus/shared-types');
var SchemaSerializer = require('./serializer/SchemaSerializer.cjs');
var NoneSerializer = require('./serializer/NoneSerializer.cjs');
var Serializer = require('./serializer/Serializer.cjs');
var schema = require('@colyseus/schema');

Serializer.registerSerializer('schema', SchemaSerializer.SchemaSerializer);
Serializer.registerSerializer('none', NoneSerializer.NoneSerializer);

exports.Client = Client.Client;
exports.ColyseusSDK = Client.ColyseusSDK;
exports.Room = Room.Room;
exports.Auth = Auth.Auth;
exports.AbortError = Errors.AbortError;
exports.MatchMakeError = Errors.MatchMakeError;
exports.ServerError = Errors.ServerError;
Object.defineProperty(exports, "CloseCode", {
	enumerable: true,
	get: function () { return sharedTypes.CloseCode; }
});
Object.defineProperty(exports, "ErrorCode", {
	enumerable: true,
	get: function () { return sharedTypes.ErrorCode; }
});
Object.defineProperty(exports, "Protocol", {
	enumerable: true,
	get: function () { return sharedTypes.Protocol; }
});
exports.SchemaSerializer = SchemaSerializer.SchemaSerializer;
exports.getStateCallbacks = SchemaSerializer.getStateCallbacks;
exports.registerSerializer = Serializer.registerSerializer;
Object.defineProperty(exports, "Callbacks", {
	enumerable: true,
	get: function () { return schema.Callbacks; }
});
